modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -347,6 +373,82 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
 
         self._info("Training complete!")
 
+    def prepare_data(self, batch: tuple) -> tuple[list, torch.Tensor, Union[torch.Tensor, dict]]:
+        sample_ids = batch[0]
+        if isinstance(sample_ids, torch.Tensor):
+            sample_ids = sample_ids.tolist()
+        elif isinstance(sample_ids, tuple):
+            sample_ids = list(sample_ids)
+
+        assert isinstance(sample_ids, list), "Cannot parse result from DataLoader"
+
+        if self._label_tranformer_function is None:
+            target = batch[2].to(self._device)
+        else:
+            target = self._label_tranformer_function(batch[2]).to(self._device)
+
+        data: Union[torch.Tensor, dict]
+        if isinstance(batch[1], torch.Tensor):
+            data = batch[1].to(self._device)
+        elif isinstance(batch[1], dict):
+            data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
+            for name, tensor in batch[1].items():
+                data[name] = tensor.to(self._device)
+        else:
+            raise ValueError(
+                "The format of the data provided is not supported in modyn. "
+                "Please use either torch tensors or dict[str, torch.Tensor]"
+            )
+
+        return sample_ids, target, data
+
+    def sample_data(self) -> None:
+        """
+        Function to score every datapoint in the current dataset and sample a fraction of it
+        Used for downsampling strategies in sample_then_batch mode
+
+        """
+        assert self._downsampler is not None
+        assert self._downsampling_mode == DownsamplingMode.SAMPLE_THEN_BATCH
+
+        # keys must be taken from the selector.
+        # This operation is needed only when we sample several times (otherwise the source is already the selector)
+        self._train_dataloader.dataset.switch_to_selector_key_source()
+
+        number_of_batches = 0
+
+        self._downsampler.init_downsampler()
+
+        sample_ids_order = []
+
+        for batch in self._train_dataloader:
+            number_of_batches += 1
+            sample_ids, target, data = self.prepare_data(batch)
+            sample_ids_order += sample_ids
+
+            with torch.autocast(self._device_type, enabled=self._amp):
+                # compute the scores and accumulate them
+                model_output = self._model.model(data)
+                self._downsampler.inform_samples(model_output, target)
+
+        downsampled_idxs, weights = self._downsampler.select_points()
+
+        selected_ids = [sample_ids_order[sample] for sample in downsampled_idxs]
+
+        # to store all the selected (sample, weight).
+        file_size = self._num_dataloaders * self._batch_size  # should we add it to the pipeline?
```

Comment:
What exactly do you mean here by adding it to the pipeline?

Reply:
Being able to specify the file_size in the pipeline

Reply:
Ah, you mean hardcoding a file size? What do you think makes sense here? Having this calculation or something like a min file size or forced file size? I am not really opinionated and would be happy for a suggestion from your side

Reply:
Ok so this is still open. I am still happy for a suggestion here, although (so we can get this merged) I think writing an issue which clearly describes what can be made configurable here and linking the issue in a `#TODO` should be fine.

Reply:
I had missed it in the other comments. I don't have an opinion honestly, it would be testing which configuration takes the least but I think it's a superfluous job. Shall I open an issue then?